.github/workflows/test.yml

```diff
@@ -1,13 +1,68 @@
-name: Unit Tests
+name: CI Pipeline
 
 on:
   pull_request:
-    branches: [ main, dataset-generator ]
+    branches: [main, dataset-generator]
+  push:
+    branches: [main, dataset-generator]
+    paths:
+      - 'rre-dataset-generator/**/*.py'
+      - '!rre-dataset-generator/**/__pycache__/**'
 
 jobs:
-  test:
+  type-check:
+    name: Type Checking (mypy)
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v4
+
+    - name: Set up Python
+      uses: actions/setup-python@v4
+      with:
+        python-version: "3.11"
+
+    - name: Install uv
+      uses: astral-sh/setup-uv@v3
+      with:
+        version: "latest"
+
+    - name: Install dependencies with dev group
+      run: uv sync --group dev
+      working-directory: rre-dataset-generator
```

Comment:
This code is common to all three jobs.
Have you considered the idea of having a single job with the following steps?
- Checkout
- Set up Python
- Install uv
- Install dependencies
- Run MyPy
- Run Linting (ruff)
- Run tests
The solution you proposed is clearer, and it's easier to identify which check is failing.
On the other side, the solution I described executes the first 4 steps only once